feat(api): Return a defined `tipPhysicallyMissing` error from `pickUpTip` commands

SyntaxColoring opened the pull request:
# Overview

This implements our first real-world instance of a Protocol Engine command returning a defined error, instead of raising an undefined error. Closes EXEC-427.

# Changelog

* Define a `tipPhysicallyMissing` error and make the `pickUpTip` command return it, when appropriate.
* Some small remaining Protocol Engine updates to handle errors that are returned in addition to errors that are raised.

# Test plan

* [x] Run a protocol with some missing tips. When it fails, the command should fail with the new `tipPhysicallyMissing` error.
    * Confirmed by looking at the raw HTTP responses.
    * The Flex ODD, however, is still showing the underlying (wrapped) error. I suspect it's considering it higher-priority because its code is numerically higher. I'm going to look at this in a separate PR. EXEC-454
* [x] Make sure the new `tipPhysicallyMissing` error shows up in the `pickUpTip` command's OpenAPI spec.
    * This *sort of* works. It does show up as a new kind of `error` value, separate from the default `ErrorOccurrence`. But it's not showing the class docstring, which is where I've tried to document the error's semantics. So we'll need to figure that out. I'd like to upgrade to Pydantic v2 before worrying about this any further.

# Review requests

See my inline GitHub comments.

# Risk assessment

Low.

SyntaxColoring committed:
WIP: Return a defined tipPhysicallyMissing error from pickUpTip.

SyntaxColoring committed:
Pass ModelUtils to all implementations.

SyntaxColoring committed:
Delete CommandT TypeVar export.

This was unused. Also, in general, sharing TypeVars from a centralized module can be a confusing thing to do, because the TypeVar isn't evaluated in the context of that module.

SyntaxColoring committed:
Delete defunct __all__ entry.

sfoster1 commented on api/src/opentrons/protocol_engine/commands/pick_up_tip.py:
```diff
@@ -50,25 +61,61 @@ class PickUpTipResult(DestinationPositionResult):
+    labware_id: str
+    well_name: str
+
+
```
this is kind of a nit, but what we do here is going to be what devs pattern match when adding things to subsequent commands. Can we establish the pattern of
- create an alias for the return type: `PickUpTipReturn = Union[SuccessData[...], DefinedErrorData[...], ...]`
- use that alias in the implementation generic type argument
- use that alias as the return annotation

SyntaxColoring committed:
Add PickUpTipImplementation unit tests.

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/pick_up_tip.py:
```diff
@@ -50,25 +61,61 @@ class PickUpTipResult(DestinationPositionResult):
+    detail: str = "No tip detected."
+
+
+@dataclass
+class TipPhysicallyMissingErrorInternalData:
+    """Internal-to-ProtocolEngine data about a TipPhysicallyMissingError."""
+
+    pipette_id: str
+    labware_id: str
+    well_name: str
```
`tips.py` will need this info to logically remove the tip from the tip rack. It's currently getting this info from the command's `params`, which was [quick hack](https://github.com/Opentrons/opentrons/blob/76360e51dd263e1a11e2f5f7168e549bc8c132ac/api/src/opentrons/protocol_engine/actions/actions.py#L175-L180) that probably won't scale to other error types.

For the sake of being incremental, I'll undo that hack in a separate PR, not this one. So, as of this PR, this data is just in anticipation of that, and is not currently consumed by anything.

SyntaxColoring committed:
Flesh out tipPhysicallyMissing semantics.

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/pick_up_tip.py:
```diff
@@ -50,25 +61,65 @@ class PickUpTipResult(DestinationPositionResult):
     )
 
 
+class TipPhysicallyMissingError(ErrorOccurrence):
+    """Returned when sensors determine that no tip was physically picked up.
+
+    That space in the tip rack is marked internally as not having any tip,
+    as if the tip were consumed by a pickup.
+
+    The pipette will act as if no tip was picked up. So, you won't be able to aspirate
+    anything, and movement commands will assume there is no tip hanging off the bottom
+    of the pipette.
+    """
```
> movement commands will assume there is no tip hanging off the bottom of the pipette.

Is this a good idea? Theoretically, we could assume that there *is* a tip hanging off the bottom of the pipette.

That might be a bit harder to implement and maintain, since there "is a tip" for movement purposes but "there is not a tip" for liquid handling purposes. But it would be safer for the client. Imagine this error triggers falsely because of a flaky sensor, so physically, there actually is a tip attached. Then, if a client commands a `moveToWell` as part of its recovery flow, we probably want the robot to go high enough to clear all the labware on the deck, right?

SyntaxColoring replied:
On the other hand, if a client commands a `moveToWell` to be followed by a `pickUpTipInPlace`, we want the robot to put the pipette nozzle in the proper tip pickup position, which means acting as if there is no tip attached...

SyntaxColoring committed:
Resolve todo comments.

The `wrappedErrors` thing looks correct.

The `errorType` thing will be part of EXEC-453.

The `detail` thing, I think yes it can have punctuation, because it can be up to "a couple of sentences."

SyntaxColoring committed:
Delete apparently unused (?) unexpected_error arg in unit test.

SyntaxColoring committed:
Fix test.

SyntaxColoring committed:
Add CommandExecutor unit test for returned errors.

SyntaxColoring committed:
Deduplicate return type union.

SyntaxColoring committed:
Merge branch 'edge' into return_error_from_pick_up_tip

An unrelated robot-server test is failing. Does this fix it?

TamarZanzouri commented on api/src/opentrons/protocol_engine/actions/actions.py:
```diff
@@ -158,13 +163,25 @@ class FailCommandAction:
 
     Must be unique to this occurrence of the error.
+
+    todo(mm, 2024-05-13): This is redundant with `error` when it's a defined error.
```
sorry for the question but is this part of the docstring?

SyntaxColoring replied:
I figured, if you're in a different file and hovering over this field in your IDE trying to figure out what the difference is between it and `error.public.id`, this would be helpful to see.

TamarZanzouri commented on api/src/opentrons/protocol_engine/actions/actions.py:
```diff
@@ -158,13 +163,25 @@ class FailCommandAction:
-    """When the command failed."""
+    """When the command failed.
+
+    todo(mm, 2024-05-13): This is redundant with `error` when it's a defined error.
```
same

SyntaxColoring commented:
Robot server test failures are unrelated, caused by AUTH-401.

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/pick_up_tip.py:
```diff
@@ -82,24 +124,46 @@ async def execute(
+        except TipNotAttachedError as e:
+            return DefinedErrorData(
+                public=TipPhysicallyMissingError(
+                    id=self._model_utils.generate_id(),
```
maybe a stupid question but why do we need 2 different id's?

SyntaxColoring replied:
Not a stupid question. One error wraps another error, and each error “has a unique ID.” I interpret that as being unique to the individual error object, not unique to the overall error event. I’m not 100% sure this is correct.

Separately: the error `id` field is useless these days, as far as I can tell, and I would like to delete it.

TamarZanzouri replied:
Thank you! yeah that was another question I had. I didn't see we use the `id` anywhere so I wondered why we have it. I guess I just figured that the wrapped error already had an id.

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/save_position.py:
```diff
@@ -53,11 +53,11 @@ class SavePositionImplementation(
         self,
         gantry_mover: GantryMover,
-        model_utils: Optional[ModelUtils] = None,
+        model_utils: ModelUtils,
```
why do we need this change?

SyntaxColoring replied:
It’s not necessary. But this is how dependency injection works in general for command implementations, and I didn’t see a reason for `ModelUtils` to be different.

TamarZanzouri reviewed (approved):
Looks great!!! just a couple of questions but overall I like this change

SyntaxColoring committed:
Merge branch 'edge' into return_error_from_pick_up_tip

Merging to fix robot-server test failures.
